src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonIEnumerableConverterFactory.cs

```diff
@@ -0,0 +1,234 @@
+                elementType = actualTypeToConvert.GetGenericArguments()[0];
+            }
+            // ISet<>
+            else if ((actualTypeToConvert = typeToConvert.GetCompatibleGenericInterface(typeof(ISet<>))) != null)
+            {
+                converterType = typeof(JsonISetOfTConverter<,>);
+                elementType = actualTypeToConvert.GetGenericArguments()[0];
+            }
+            // ICollection<>
+            else if ((actualTypeToConvert = typeToConvert.GetCompatibleGenericInterface(typeof(ICollection<>))) != null)
+            {
+                converterType = typeof(JsonICollectionOfTConverter<,>);
+                elementType = actualTypeToConvert.GetGenericArguments()[0];
+            }
+            // Stack<> or deriving from Stack<>
+            else if ((actualTypeToConvert = typeToConvert.GetCompatibleGenericBaseClass(typeof(Stack<>))) != null)
+            {
+                converterType = typeof(JsonStackOfTConverter<,>);
+                elementType = actualTypeToConvert.GetGenericArguments()[0];
+            }
+            // Queue<> or deriving from Queue<>
+            else if ((actualTypeToConvert = typeToConvert.GetCompatibleGenericBaseClass(typeof(Queue<>))) != null)
+            {
+                converterType = typeof(JsonQueueOfTConverter<,>);
+                elementType = actualTypeToConvert.GetGenericArguments()[0];
+            }
+            // ConcurrentStack<> or deriving from ConcurrentStack<>
+            else if ((actualTypeToConvert = typeToConvert.GetCompatibleGenericBaseClass(typeof(ConcurrentStack<>))) != null)
+            {
+                converterType = typeof(JsonConcurrentStackOfTConverter<,>);
+                elementType = actualTypeToConvert.GetGenericArguments()[0];
+            }
+            // ConcurrentQueue<> or deriving from ConcurrentQueue<>
+            else if ((actualTypeToConvert = typeToConvert.GetCompatibleGenericBaseClass(typeof(ConcurrentQueue<>))) != null)
+            {
+                converterType = typeof(JsonConcurrentQueueOfTConverter<,>);
+                elementType = actualTypeToConvert.GetGenericArguments()[0];
+            }
+            // IEnumerable<>, types assignable from List<>
+            else if ((actualTypeToConvert = typeToConvert.GetCompatibleGenericInterface(typeof(IEnumerable<>))) != null)
+            {
+                converterType = typeof(JsonIEnumerableOfTConverter<,>);
+                elementType = actualTypeToConvert.GetGenericArguments()[0];
+            }
+            // Check for non-generics after checking for generics.
+            else if (typeof(IDictionary).IsAssignableFrom(typeToConvert))
+            {
+                if (typeToConvert == typeof(IDictionary))
+                {
+                    return s_IDictionaryConverter;
+                }
+
+                converterType = typeof(JsonIDictionaryConverter<>);
+            }
+            else if (typeof(IList).IsAssignableFrom(typeToConvert))
```

Comment:
Since dictionary is a "higher level" concept than list, dictionary should be first. Consider if someone wrote a collection that derived from both IList and IDictionary -- I'd except IDictionary to "win".

Reply:
I see. That makes sense.